### ISSUE TITLE
TestUGen_RTAlloc:test_allUGens relies on being able to fill the server's realtime memory (probably source of CI bugs)

### ISSUE THREAD

**Issue by JordanHendersonMusic** [supercollider/supercollider#6410]
## Description

This test assumes the server's real time memory is exactly as stated and does include any padding.

## Environment
* SuperCollider version: current

## Explanation

Once you unpack the code it boils down to the following (removing synchronisation to make things clearer)...
```supercollider
{ AllpassC.ar(DC.ar(1), 1, 0) }.loadToFloatArray(0.001, s) { |out|
	out.sum != 0  << first test
};

{ LocalBuf(1024 * 1000 / 4) }.play;

{ AllpassC.ar(DC.ar(1), 1, 0)  }.loadToFloatArray(0.001, s) { |out|
	out.sum == 0  << second test
};
```

I believe the point here is to fill the server's real time memory with the buffer so the allpass allocationfails.

There are several things wrong with this.
1. The server's memory does not exactly have to be as reported, in can be more.
2. The `LocalBuf` UGen doesn't have any side effects and would be completely reasonable for the synthdef compiler to remove it #6405.

**Comment by JordanHendersonMusic** [comment]
@capital-G do you think my reading of the code is correct?

**Comment by capital-G** [comment]
The tests are indeed a bit opaque, e.g. what is this syntax? A method call followed by a function?

https://github.com/supercollider/supercollider/blob/d52b0522b210def4fe47829ca332f0bc411b3509/testsuite/classlibrary/TestUGen_RTAlloc.sc#L166-L170

Maybe I should go back to SC Syntax school :D

Nonetheless, I'll try to put it into my words from what I understand by repeating.

IIUC the idea is to spawn stateful UGens which allocate memory - first time with enough memory available, the 2nd time w/o enough memory.

By using delaytime values such as `0.0` they should simply bypass the signal the original signal in this case, so if we have `DC.ar(1.0)` as input, the sum of the rendered/returned float array should be > 0.

In the 2nd case we allocate some memory on the server, with the means of an `LocalBuf` which consumes the whole memory of the server.
If we now spawn such a stateful UGen, it should fail (because of mem alloc error) and the UGen should return 0.0.

I tried to reproduce this locally, but I seem to have some wrong calculations concerning RT memory?

```supercollider

// the setup
{DC.ar(0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})
{DC.ar(1.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})

// now with delay
{DelayC.ar(DC.ar(1.0), maxdelaytime: 1.0, delaytime: 0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})

// ok


// max num of floats we can allocate on the server - turn kbytes into bits and divide it by 32bit float
~maxNumFloats = s.options.memSize * 8 * 1000 / 32;

// consume all of the RT memory - this time with an empty delay line
// via a BufDelay so there would be side effects
(
{
	var buf = LocalBuf(numFrames: ~maxNumFloats, numChannels: 1);
	BufDelayN.ar(buf, DC.ar(0.0));
}.play;
)

// this does not fail?!
{DelayC.ar(DC.ar(1.0), maxdelaytime: 0.2, delaytime: 0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})

// this does fail !?
{DelayC.ar(DC.ar(1.0), maxdelaytime: 1.0, delaytime: 0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})

// where is my math wrong?!
```

In general, think the removal of the allocation due to its lacking side effects would be valid in this case (or maybe enable it via a default flag?).

**Comment by JordanHendersonMusic** [comment]
```
// this does not fail?!
{DelayC.ar(DC.ar(1.0), maxdelaytime: 0.2, delaytime: 0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})

// this does fail !?
{DelayC.ar(DC.ar(1.0), maxdelaytime: 1.0, delaytime: 0.0)}.loadToFloatArray(duration: 0.01, action: {|a| a.sum.postln;})
```
In between these two lines is the creation of a large localbuf. I think it is trying to reset the server's memory.

**Comment by JordanHendersonMusic** [comment]
Ah you're right, I've read it wrong! This issue is named wrong then.

So the problem is that the server isn't out of memory, not that the UGens are wrong, and that we are testing that the reported memory size is exactly as stated and doesn't include any padding.

Seems like this should be accomplished by a new server command that just disables rt allocation.

**Comment by JordanHendersonMusic** [comment]
> what is this syntax? A method call followed by a function?

That's trailing block/closure syntax... easier to demonstrate...
```
if(a) { b } == if(a, { b })
```

**Comment by capital-G** [comment]
> > what is this syntax? A method call followed by a function?
> 
> That's trailing block/closure syntax... easier to demonstrate...
> 
> ```
> if(a) { b } == if(a, { b })
> ```

Thanks. Wow, I always thought this was somehow a primitive of `if` and not generally available - still not the biggest fan of this syntax though, especially beyond control structures. 😕 

> Seems like this should be accomplished by a new server command that just disables rt allocation.

I am not sure - I think the idea is to test also the memory allocation of the server, which seems to be not working as intended?
Turning of the allocation would most likely fix the test, but I think the test is actually demonstrating that something is off?

```supercollider
s.options.memSize
// ->8192

// max num of floats - turn kbytes into bits and divide it by 32bit float
~maxNumFloats = s.options.memSize * 8 * 1000 / 32;

// consume all of the RT memory AND some more beyond without failing?!
(
{
	var buf = LocalBuf(numFrames: ~maxNumFloats+43191, numChannels: 1);
	BufDelayN.ar(buf, DC.ar(0.0));
}.play;
)

// consume all of the RT memory AND some more beyond, but now it fails ?!
(
{
	var buf = LocalBuf(numFrames: ~maxNumFloats+43192, numChannels: 1);
	BufDelayN.ar(buf, DC.ar(0.0));
}.play;
)
```

I tried to find the actual rt_malloc implemntation and I think it must be somewhere around https://github.com/supercollider/supercollider/blob/d52b0522b210def4fe47829ca332f0bc411b3509/common/SC_AllocPool.cpp#L283  - but I think this goes really beyond my expertise :/

I also managed to "crash the server" in not so nice ways while allocating memory. If you just allocate the right amount of floats you can get this message:

```
exception in real time: PerformCompletionMsg allocation failed: out of memory!
```

It is hard to reproduce, but maybe this happens when one is exactly allocating all the memory but there is no space left to allocate an OSC message? One would need to be really lucky to get the right numbers here.

**Comment by capital-G** [comment]
@elgiano as you were a person that did some work on the alloc pool and especially the alloc fail implementation (see https://github.com/supercollider/supercollider/commit/7f0804cfecda3f0ca65be37cd2a7b8a09669f333), do you have a clue what could the problem here? Especially regrading the ability to allocate more memory than declared as the example in the post above demonstrates?

**Comment by JordanHendersonMusic** [comment]
Just to throw a spanner into the works.... over here #6405 I've disable the optimisations for the memory filling synth and this test seems to pass every single time. Perhaps this is a UGen ordering issue? It is worth nothing that currently the ordering under certain circumstances isn't repeatable, explaining why there might be inconsistencies.

**Comment by dyfer** [comment]
@elgiano since you originally added these tests as part of [RTAlloc rework](https://github.com/supercollider/supercollider/pull/5713), do you have any insight why they might be failing somewhat inconsistently? CI tests are in general a little bit more reliable these days than back when we used QPM (that's a subjective observation...), thought they are not bulletproof for sure.

**Comment by capital-G** [comment]
I disabled the test in https://github.com/supercollider/supercollider/pull/6429 for now in order to have a clean CI again - would be great to not ignore this tests though, so any help to get them working/passing again would be appreciated.

**Comment by elgiano** [comment]
Hey everyone, thanks for looking at this and sorry if the code I wrote wasn't that immediate to understand!
You are right: the idea is to fill RT memory and test failed allocations, if you're curious you can check this comment I made at the time https://github.com/supercollider/supercollider/pull/5713#issuecomment-1056171574

I'm looking at it, but first I want to reply to @capital-G above. Your strange alloc results could be due to a calculation error:
```
// max num of floats - turn kbytes into bits and divide it by 32bit float
~maxNumFloats = s.options.memSize * 8 * 1000 / 32;

// doesn't fail
x = { BufDelayN.ar(LocalBuf(~maxNumFloats, 1), DC.ar(0.0)) }.play;
x.free

// max num of floats - turn kbytes into bytes and divide it by 4bytes float
~maxNumFloats = s.options.memSize * 1024 / 4

// fails
{ BufDelayN.ar(LocalBuf(~maxNumFloats, 1), DC.ar(0.0)) }.play;
```
I'm also not 100% sure, but you can see memSize being multiplied by 1024 before being passed to AllocPool constructor in [scsynth_main.cpp.](https://github.com/supercollider/supercollider/blob/71d5d1841bc1b26a59560c6544f3996765f6ff63/server/scsynth/SC_World.cpp#L328)

**Comment by elgiano** [comment]
About the tests:

This test has problems, indeed, failing more or less unpredictably (on my machine it always passes, obviously). I'm looking at old CI logs and so far found two scenarios. Reports are below.

**tl;dr**: sometimes LocalBuf allocation fails, making all allocFail tests fail. Might be due to other real-time memory being allocated at that time, preventing us from allocating the whole of it. Plus if I understand correctly, with new proposed changes LocalBuf could be optimized out if not used (it wasn't the case before). So, I'm considering re-writing these tests without it, like @capital-G said:
> One pragmatic way would be to simply spawn the UGens with an obscene amount of memory in order to "crash" them, although I am unsure if this is feasible for all UGens.

I'm going through it, will let you know as soon as I can <3

P.S. @JordanHendersonMusic if you have it fresh, could you point me out to when can the server allocate more RT memory than specified? I didn't know it could [shy emoji]. Would it be appropriate to write it here? Otherwise a DM on scsynth would make me as happy. Thanks!!

## Reports
### all allocFail tests fail
It seems like allocFail tests fail because LocalBuf allocation fails:
```
PASS: a TestUGen_RTAlloc: test_allUGens - Limiter allocPass test should produce non-zero output

LocalBuf_allocBuffer: alloc failed, increase server's RT memory (e.g. via ServerOptions)

FAIL: a TestUGen_RTAlloc: test_allUGens - Convolution allocFail test should produce zero output
FAIL: a TestUGen_RTAlloc: test_allUGens - DelayL allocFail test should produce zero output
FAIL: a TestUGen_RTAlloc: test_allUGens - AllpassC allocFail test should produce zero output
FAIL: a TestUGen_RTAlloc: test_allUGens - Pitch allocFail test should produce zero output
```

My interpretation is that some other "process" (OSC messages? residuals from previous SynthDefs? I'll try to find out) is allocating memory, so that LocalBuf can't allocate all of it, and it ends up allocating none, thus making all subsequent allocFail tests fail.  At this point I assume the test only passed because of an error I made calculating LocalBuf's size, effectively resulting in a bit less than s.options.memSize.

Here are the CI runs I inspected:
https://github.com/supercollider/supercollider/actions/runs/10117988847/job/27984106431
https://github.com/supercollider/supercollider/actions/runs/10126536884/job/28013381641
https://github.com/supercollider/supercollider/actions/runs/10110787893/job/27961525058 
https://github.com/supercollider/supercollider/actions/runs/10085292685/job/27885925002

### MFCC allocPass fail
I need to investigate this more, but I guess MFCC is correctly outputting all zeros, tricking the test.  Giving it more time to run (than 0.001) would fix it. It's however hard to test this UGen, as we can't make it allocate as much memory as we want (max 84 floats). So, if we want to keep this allocPass test, it will become a special case.

https://github.com/supercollider/supercollider/actions/runs/10118004832/job/27984402193
https://github.com/supercollider/supercollider/actions/runs/10127415823/job/28005706193
https://github.com/supercollider/supercollider/actions/runs/10067337695/job/27831392915 

This

**Comment by JordanHendersonMusic** [comment]
Hi @elgiano 😊 

> could you point me out to when can the server allocate more RT memory than specified?

I misspoke here, it's not that it does this, but if it did, it would not be an error - I'd still consider it a part of the interface if it didn't allocate exactly the amount of memory specified (so long as it was at least that amount). I don't know if it's a good idea to depend on what the server does when you expect an error, but it is able to continue regardless. Perhaps there could be some way to disable all allocations from sclang (an undocumented debug osc message in scsynth?), or these tests could be done in c++ and a custom allocator provided at compile time? I'm not familiar with the c++ tests so this might be infeasible.

**Comment by mtmccrea** [comment]
> CmdPeriod.run;
// consume all of the RT memory AND some more beyond, but now it fails ?!

This problem may not be central to the issue at hand, but seems problematic... could it be that after first overshooting memory bounds (though not reporting?), memory isn't properly freed (or memory counter not properly reset) on cmd+., so it the failure is caught on the second attempt?

If this bad behavior is repeatable it should perhaps be filed as a separate Issue.

**Comment by elgiano** [comment]
@mtmccrea
I think it's worth opening another issue to investigate this, and it would be nice to have some instruments to monitor memory usage.  UGens should not leak memory when failing allocation (it used to be so before the RTAlloc rework mentioned above), but we need some tools to monitor this.

Sclang has `totalFree` for its own AllocPool, not real time memory. Scsynth has `World_TotalFree`, but it's not used anywhere... and it doesn't look difficult to make a new server command that replies this info, so that we can monitor memory usage across calls.

**Comment by capital-G** [comment]
> Sclang has `totalFree` for its own AllocPool, not real time memory. Scsynth has `World_TotalFree`, but it's not used anywhere... and it doesn't look difficult to make a new server command that replies this info, so that we can monitor memory usage across calls.

Something like this https://github.com/supercollider/supercollider/compare/develop...capital-G:supercollider:rtmem-server-command @elgiano ?  Makes me even more confused - but was a long day, maybe I oversee something obvious here...

But there seems to be something like a memory leak if you exceed the allocation?

```supercollider
s.options.memSize;
// -> 8192
s.boot;


// [0] availableMem, [1] largestChunk
OSCdef(\memPrinter, {|m| m.postln}, "/rtmem.reply", recvPort: 57120)

~rtMem = {NetAddr("127.0.0.1", 57110).sendMsg("/rtmem")}

~rtMem.();
// [/rtmem.reply, 8369024, 8369024]

x = {DelayN.ar(DC.ar(0.0),maxdelaytime: 4.0)}.play;
~rtMem.();
// [/rtmem.reply, 7317760, 7316736]

x.free;
~rtMem.();
// [/rtmem.reply, 8369024, 8369024]

// assuming 48khz samplerate
x = {DelayN.ar(DC.ar(0.0), maxdelaytime: 1000*s.options.memSize/4/48000)}.play;
// DelayN_Ctor: alloc failed, increase server's RT memory (e.g. via ServerOptions)

~rtMem.();
// [/rtmem.reply, 8366208, 8365376]
// now less memory as 8366208 < 8369024 ?!
// and largest chunk is smaller?!

// why /2?! is .play doing some implicit multi channel expansion here?
x = {DelayN.ar(DC.ar(0.0), maxdelaytime: (s.options.memSize * 1000 / 4 / 48000) / 2.0)}.play;
~rtMem.();
// [/rtmem.reply, 4172032, 4171008]

CmdPeriod.run();
~rtMem.();
// [/rtmem.reply, 8369024, 8369024]
```

**Comment by elgiano** [comment]
@capital-G 
It's the delay allocation itself that likely doubles the memory allocation:
```
mem = nextPowerOfTwo(maxdelaytime * s.sampleRate + 1 + s.options.blockSize)
```

https://github.com/supercollider/supercollider/blob/e49e42cbf42be6dc9ff4806cf14ff9b9d14e0bb5/server/plugins/DelayUGens.cpp#L3440C1-L3449C1
```cpp
static bool DelayUnit_AllocDelayLine(DelayUnit* unit) {
    long delaybufsize = (long)ceil(unit->m_maxdelaytime * SAMPLERATE + 1.f);
    delaybufsize = delaybufsize + BUFLENGTH;
    delaybufsize = NEXTPOWEROFTWO(delaybufsize); // round up to next power of two
    unit->m_fdelaylen = unit->m_idelaylen = delaybufsize;


    if (unit->m_dlybuf)
        RTFree(unit->mWorld, unit->m_dlybuf);
    unit->m_dlybuf = (float*)RTAlloc(unit->mWorld, delaybufsize * sizeof(float));

```

**Comment by elgiano** [comment]
Since it seems interesting, and perhaps this issue will be closed soon, I made a PR for the rtMemStatus code #6467

**Comment by mtmccrea** [comment]
@capital-G 
>edit: also got [/rtmem.reply, 8369024, 8368832] after a "cold boot"

did you discover what was going on with this issue?
